File: sdk/samples/all in on code/Visual Studio 2008/VBWinFormDesigner/PropertyDisplayOrder/UC_PropertyDisplayOrder.vb
'************************************* Module Header **************************************\
' Module Name:	UC_PropertyDisplayOrder.vb
' Project:		VBWinFormDesigner
' Copyright (c) Microsoft Corporation.
' 
' 
' The PropertyDisplayOrder sample demonstrates how to change the default display order on 
' Properties windows for properties, and how to expand a nested property on the Properties 
' window. 
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
'
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************'

Imports System.ComponentModel

Namespace PropertyDisplayOrder

    Public Class UC_PropertyDisplayOrder
        Private _cls As SubClass = New SubClass

        <TypeConverter(GetType(PropertyOrderConverter))> _
    Public Property Cls() As SubClass
            Get
                Return Me._cls
            End Get
            Set(ByVal value As SubClass)
                Me._cls = value
            End Set
        End Property
    End Class

    Public Class SubClass
        Private _prop1 As String
        Private _prop2 As Integer
        Private _prop3 As Single

        Public Property Prop1() As String
            Get
                Return Me._prop1
            End Get
            Set(ByVal value As String)
                Me._prop1 = value
            End Set
        End Property

        Public Property Prop2() As Integer
            Get
                Return Me._prop2
            End Get
            Set(ByVal value As Integer)
                Me._prop2 = value
            End Set
        End Property

        Public Property Prop3() As Single
            Get
                Return Me._prop3
            End Get
            Set(ByVal value As Single)
                Me._prop3 = value
            End Set
        End Property
    End Class


    Public Class PropertyOrderConverter
        Inherits ExpandableObjectConverter

        Public Overrides Function GetProperties(ByVal context As ITypeDescriptorContext, _
                    ByVal value As Object, ByVal attributes As Attribute()) As PropertyDescriptorCollection
            Dim propertyDescriptorCollection As PropertyDescriptorCollection = _
            TypeDescriptor.GetProperties(GetType(SubClass), attributes)
            Dim propNames As String() = New String() {"Prop2", "Prop3", "Prop1"}
            Return propertyDescriptorCollection.Sort(propNames)
        End Function

        Public Overrides Function GetPropertiesSupported(ByVal context As ITypeDescriptorContext) As Boolean
            Return True
        End Function

    End Class


End Namespace

File: sdk/samples/all in on code/Visual Studio 2008/CSWinFormDesigner/PropertyDisplayOrder/UC_PropertyDisplayOrder.cs
/************************************* Module Header **************************************\
* Module Name:	UC_PropertyDisplayOrder.cs
* Project:		CSWinFormDesigner
* Copyright (c) Microsoft Corporation.
* 
* 
* The PropertyDisplayOrder sample demonstrates how to change the default display order on 
* Properties windows for properties, and how to expand a nested property on the Properties 
* window. 
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* History:
* * 5/25/2009 3:00 PM Zhi-Xin Ye Created
* * 
* 
\******************************************************************************************/

#region namespaces
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
#endregion

namespace CSWinFormDesigner.PropertyDisplayOrder
{
    public partial class UC_PropertyDisplayOrder : UserControl
    {
        public UC_PropertyDisplayOrder()
        {
            InitializeComponent();
        }

        private SubClass cls = new SubClass();

        [TypeConverter(typeof(PropertyOrderConverter))]
        public SubClass Cls
        {
            get { return this.cls; }
            set { this.cls = value; }
        }
    }

    public class SubClass
    {
        private string prop1;
        private int prop2;
        private float prop3;

        public string Prop1
        {
            get { return prop1; }
            set { prop1 = value; }
        }
        public int Prop2
        {
            get { return prop2; }
            set { prop2 = value; }
        }
        public float Prop3
        {
            get { return prop3; }
            set { prop3 = value; }
        }
    }

    public class PropertyOrderConverter : ExpandableObjectConverter
    {
        /*
         * Change properties display order in Properties Window
         */
        #region
        public override bool GetPropertiesSupported(ITypeDescriptorContext context)
        {
            return true;
        }

        public override PropertyDescriptorCollection GetProperties(ITypeDescriptorContext context, object value, Attribute[] attributes)
        {
            System.ComponentModel.PropertyDescriptorCollection propertyDescriptorCollection;

            string[] propNames;

            propertyDescriptorCollection =
                TypeDescriptor.GetProperties(typeof(SubClass), attributes);

            propNames = (string[])new System.String[3];

            //Specify the order for the properties
            propNames[0] = @"Prop2";
            propNames[1] = @"Prop3";
            propNames[2] = @"Prop1";

            return propertyDescriptorCollection.Sort(propNames);
        }
        #endregion
    }
}
